File: packages/core/src/dna/gene-extractors/responsive-approach.ts
import { BaseGeneExtractor, type AlleleDefinition, type FileExtractionResult, type DetectedAllele } from './base-extractor.js';

import type { GeneId } from '../types.js';

export class ResponsiveApproachExtractor extends BaseGeneExtractor {
  readonly geneId: GeneId = 'responsive-approach';
  readonly geneName = 'Responsive Approach';
  readonly geneDescription = 'How responsive design is handled';

  getAlleleDefinitions(): AlleleDefinition[] {
    return [
      { id: 'tailwind-mobile-first', name: 'Tailwind Mobile-First', description: 'sm:, md:, lg: breakpoints', patterns: [/className\s*=\s*["'`][^"'`]*\b(sm|md|lg|xl):/] },
      { id: 'tailwind-desktop-first', name: 'Tailwind Desktop-First', description: 'max-* variants', patterns: [/className\s*=\s*["'`][^"'`]*\bmax-(sm|md|lg):/] },
      { id: 'css-media-queries', name: 'CSS Media Queries', description: '@media queries', patterns: [/@media\s*\([^)]*width/] },
      { id: 'container-queries', name: 'Container Queries', description: '@container', patterns: [/@container/] },
      { id: 'js-responsive', name: 'JS Responsive', description: 'useMediaQuery hooks', patterns: [/useMediaQuery\s*\(/, /useBreakpoint\s*\(/] },
    ];
  }

  extractFromFile(fp: string, c: string, _imp: string[]): FileExtractionResult {
    const det: DetectedAllele[] = [];
    for (const d of this.getAlleleDefinitions()) {
      for (const p of d.patterns) { const m = p.exec(c); if (m) { const ctx = this.extractContext(c, m.index); det.push({ alleleId: d.id, line: ctx.line, code: ctx.code, confidence: 0.8, context: ctx.context }); break; } }
    }
    return { file: fp, detectedAlleles: det, isComponent: this.isComponentFile(fp, c) };
  }
}

File: drift v2/crates/drift/drift-analysis/src/structural/dna/extractors/responsive_approach.rs
//! Frontend gene: responsive-approach — how responsive design is implemented.

use crate::structural::dna::extractor::GeneExtractor;
use crate::structural::dna::types::*;
use super::variant_handling::extract_with_definitions;

pub struct ResponsiveApproachExtractor;

impl GeneExtractor for ResponsiveApproachExtractor {
    fn gene_id(&self) -> GeneId { GeneId::ResponsiveApproach }

    fn allele_definitions(&self) -> Vec<AlleleDefinition> {
        vec![
            AlleleDefinition {
                id: "tailwind-breakpoints".into(), name: "Tailwind Breakpoints".into(),
                description: "Uses Tailwind responsive prefixes (sm:, md:, lg:)".into(),
                patterns: vec![r"\b(sm|md|lg|xl|2xl):".into()],
                keywords: vec!["sm:".into(), "md:".into(), "lg:".into()],
                import_patterns: vec![], priority: 10,
            },
            AlleleDefinition {
                id: "media-queries".into(), name: "CSS Media Queries".into(),
                description: "Uses @media queries for responsive design".into(),
                patterns: vec![r"@media\s*\(".into(), r"useMediaQuery".into()],
                keywords: vec!["@media".into()],
                import_patterns: vec![], priority: 7,
            },
            AlleleDefinition {
                id: "container-queries".into(), name: "Container Queries".into(),
                description: "Uses CSS container queries".into(),
                patterns: vec![r"@container\s*\(".into(), r"container-type".into()],
                keywords: vec!["@container".into()],
                import_patterns: vec![], priority: 8,
            },
            AlleleDefinition {
                id: "css-grid-flex".into(), name: "CSS Grid/Flexbox".into(),
                description: "Uses CSS Grid or Flexbox for responsive layouts".into(),
                patterns: vec![r"display:\s*(?:grid|flex)".into(), r"\bgrid-template".into()],
                keywords: vec!["grid".into(), "flexbox".into()],
                import_patterns: vec![], priority: 5,
            },
        ]
    }

    fn extract_from_file(&self, content: &str, file_path: &str) -> FileExtractionResult {
        extract_with_definitions(content, file_path, &self.allele_definitions())
    }
}
